---
title: "test summary report"
date: last-modified
format:
    html: default
    ##pdf: default
editor: 
    render-on-save: true
---

# so you've run a CRISPR functional screen but what does it mean?

pooled CRISPR functional screens are typically used for detecting cell gene expression in a variety of conditions (whatever you want to test, really). after your wet lab experiments and sequencing, our analysis of the raw sequences gives us counts of a particular genotype (measured by sgRNA or gene presence). this report strives to break down our analysis and show you preliminary results on the success of your experiment so you can move forward with your investigation.

# analysis summary table

we've included a summary table of how your raw sequences fared in our analysis. with a successful CRISPR functional screen, you'd expect most of the reads to contain the vector and to align to sgRNA.

```{r}
#| label: tbl-summaryStats
#| tbl-cap: "sample counts per analysis step"
#| tbl-colwidths: [60,40]

library(knitr)
library(readr)

summary_table <- read_tsv('summary_stats/summary_stat_table.tsv', show_col_types = FALSE)

new_cols <- c('Sample ID',
'Total Reads Sequenced',
'Reads with Vector',
'Reads without Vector',
'Reads Aligned to sgRNA',
'Reads Not Aligned to sgRNA',
'Reads with Alignments Suppressed',
'Total sgRNA Represented',
'sgRNA > 10 Reads')

colnames(summary_table) <- new_cols

kable(summary_table, align='lcccccccc')
```

| Column                           | Explanation                                                                                                                                                              |
|-----------------------|-------------------------------------------------|
| Total Reads Sequenced            | Total number of reads sequenced in your sample.                                                                                                                          |
| Reads with Vector                | Total number of reads containing the vector sequence. This is the number of reads used for alignment.                                                                    |
| Reads without Vector             | Total number of reads where the vector sequence was not detected and therefore not considered for sgRNA mapping. These reads are not included in the alignment step.     |
| Reads Aligned to sgRNA           | Total number of reads that contained a vector sequence that had a read aligning to a sgRNA at the set mismatch rate of {{< var bowtie_mismatches >}}.                    |
| Reads Not Aligned to sgRNA       | Total number of reads that contained a vector sequence where the vector sequence **did not** align to a sgRNA at the set mismatch rate of {{< var bowtie_mismatches >}}. |
| Reads with Alignments Suppressed | Total number of reads that contained a vector sequence that aligned more than once to a sgRNA. We want to guarantee that all reported alignments to a sgRNA are unique.  |
| Total sgRNA Represented          | The total number of unique guides with at least 1 read count detected. Percentage is based upon fasta of guide library used as the reference.                            |
| sgRNA \> 10 Reads                | The total number of unique guides with at least 10 read counts detected. Percentage is based upon fasta of guide library used as the reference.                          |

: summary table explained {tbl-colwidths="\[25,75\]".striped}

# what we do

below is a more in-depth description of how we process your raw sequences including the software used, outputs, and why we do each particular step.

note: for the workflow .png to render in a non-blurry format you'll need to save it to the exact px dimensions that you want it to be in for the html (so this file). right now it's; w: 2850 px, h: 2311 px, dpi: 688.27.

![](plots/workflow_flowchart.png){width="2850"}

| Workflow Step                            | Why we do this                                                                                                                                                                              |
|----------------------|--------------------------------------------------|
| 1: Quality Check                         | Done to assess the overall quality of the sequences we received prior to additional analysis.                                                                                               |
| 2: Raw Sequence Prep (trimming)          | To trim all found sequences from the 5' end to 20 base pairs (bp) in length based on the vector sequence of the library that was used. All sgRNA sequences should be about 20 bp in length. |
| 3: Sequence Alignment                    | To map the trimmed sgRNA sequences to the associated library that they were prepped with to see which sgRNA sequences are present.                                                          |
| 4: Counting sgRNA Sequences              | To count how many sgRNA sequences and genes mapped to the sgRNA index for each sample.                                                                                                      |
| 5: Combining sgRNA Count Files           | To combine the sgRNA count data for each sample into one file for downstream R comparison analysis.                                                                                         |
| 6: Normalizing Combined sgRNA Count File | To generate all needed files to perform PCA, total/transformed count, and correlation R analysis. sgRNA and gene count data are normalized to counts per million (CPM).                     |
| 7: PCA Analysis                          | Plotted at the sgRNA and gene level to see if biological replicates cluster together as they should.                                                                                        |
| 8: Total/Transformed Count Comparison    | Plotted at the sgRNA and gene level to ensure that biological replicates have about the same distribution of counts.                                                                        |
| 9: Correlation Comparison                | Plotted at the sgRNA and gene level to ensure that biological group replicates correlate more closely with each other than replicates in a different biological group.                      |

: workflow diagram explained {tbl-colwidths="\[25,75\]".striped}

# plots

we generate plots to validate that your experiment ran smoothly and that all biological replicates look similar to each other. there are a few different ways we do this; total and transformed sgRNA/gene counts, PCA analysis, and correlation analysis.

## total and transformed counts

we looked at total and transformed sgRNA/gene counts to make sure that all samples and biological replicates have about the same amount (there isn't a separate one for individual genes since they both add up to the same number).

### total counts

total counts is exactly what it sounds like, adding up all the sgRNA/gene counts per sample for comparison. the plot is colored by biological group and all replicates should be next to each other on the x-axis.

![Total Count Plot](plots/sample_totalCount_plot.png)

### transformed counts

the total sgRNA/gene counts above undergo a log2() transformation to ensure that the distribution of counts per sample is similar across all samples. like the plot above, the plot is colored by biological group and all replicates should be next to each other on the x-axis. **(why are we doing this??)**

![log2(Total Count Plot)](plots/sample_transCounts_plot.png)

## pca analysis

we next put the counts-per-million (CPM) normalized sgRNA and gene counts through a principal components analysis (PCA) to look at the clustering of all replicates in a biological group. we expect replicates in the same biological group to cluster together (this is not a good example). we do this for sgRNA and gene counts separately since they're different and thus give us slightly different PCA results.

### sgRNA count PCA

we did this using all of the small guide RNA (sgRNA) sequences

![sgRNA Count PCA plot](plots/sgRNACount_PCA_plot.png)

### gene count PCA

this was done using the individual genes (several sgRNAs typically fall under one gene, so these are just added up).

![Gene Count PCA plot](plots/geneCount_PCA_plot.png)

## correlation analysis

we performed a spearman correlation analysis to make sure that all replicates in a biological group correlate more closely with each other than other biological groups (i.e. we want a correlation value closer to 1). positive correlations will be deeper blue and negative correlations will be deeper red. we ran our correlation analysis on sgRNA and gene counts separately.

### sgRNA correlation comparisons

we did this for all small guide RNA (sgRNA) sequences - idk if we need this text

![sgRNA Count Correlation Plot](plots/sgRNACount_correlationMatrix.png)

### gene correlation comparisons

this was done using the individual genes - same with this text

![Gene Count Correlation Plot](plots/geneCount_correlationMatrix.png)

# contact information

If you have any questions regarding your results, our analysis, or ways to further process your data please feel free to contact us! As always, if we have done any analysis for you that ends up in a publication please consider including us in the author list.

**Madi Apgar, MS: madison.apgar\@cuanschutz.edu**\
**Tonya Brunetti, PhD: tonya.brunetti\@cuanschutz.edu**

This workflow is publically available as a [GitHub repository](https://github.com/Comp-Bio-Pipeline-Dev-Team/functional_CRISPR_screen).